src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+    }
+    mTimeSyncDataProvider.LoadTimeZone(mTimeZoneList);
+    mTimeSyncDataProvider.LoadDSTOffset(mDstOffList);
+    // TODO if trusted time source is available schedule a time read
+    if (!mTrustedTimeSource.IsNull())
+    {
+    }
+}
+
+void TimeSynchronizationServer::ScheduleDelayedAction(System::Clock::Seconds32 delay, System::TimerCompleteCallback action,
+                                                      void * aAppState)
+{
+    VerifyOrDie(SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState) ==
+                CHIP_NO_ERROR);
+}
+
+namespace {
+
+struct TimeZoneCodec
+{
+    static constexpr TLV::Tag TagOffset() { return TLV::ContextTag(TimeSynchronization::Structs::TimeZoneStruct::Fields::kOffset); }
+    static constexpr TLV::Tag TagValidAt()
+    {
+        return TLV::ContextTag(TimeSynchronization::Structs::TimeZoneStruct::Fields::kValidAt);
+    }
+    static constexpr TLV::Tag TagName() { return TLV::ContextTag(TimeSynchronization::Structs::TimeZoneStruct::Fields::kName); }
+
+    TimeSynchronization::Structs::TimeZoneStruct::Type timeZone;
+
+    TimeZoneCodec(TimeSynchronization::Structs::TimeZoneStruct::Type tz) : timeZone(tz) {}
+
+    static constexpr bool kIsFabricScoped = false;
+
+    CHIP_ERROR Encode(TLV::TLVWriter & writer, TLV::Tag tag) const
+    {
+        TLV::TLVType outer;
+        ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Structure, outer));
+
+        // Offset
+        ReturnErrorOnFailure(DataModel::Encode(writer, TagOffset(), timeZone.offset));
+        // ValidAt
+        ReturnErrorOnFailure(DataModel::Encode(writer, TagValidAt(), timeZone.validAt));
+        // Name
+        if (timeZone.name.HasValue())
+        {
+            uint32_t name_size = static_cast<uint32_t>(strnlen(timeZone.name.Value().data(), 64));
+            ReturnErrorOnFailure(writer.PutString(TagName(), timeZone.name.Value().data(), name_size));
+            ChipLogProgress(Zcl, "%s %d", timeZone.name.Value().data(), name_size);
+        }
+        ReturnErrorOnFailure(writer.EndContainer(outer));
+        return CHIP_NO_ERROR;
+    }
+};
+
+static bool computeLocalTime(chip::EndpointId ep)
+{
+    DataModel::Nullable<uint64_t> utcTime, localTime;
+    int32_t timeZoneOffset = 0, dstOffset = 0;
+    UTCTime::Get(ep, utcTime);
+    if (utcTime.IsNull())
+    {
+        return false;
+    }
+    auto tz  = TimeSynchronizationServer::Instance().GetTimeZone().begin();
+    auto dst = TimeSynchronizationServer::Instance().GetDSTOffset().begin();
+    if (tz->validAt <= utcTime.Value())
+    {
+        timeZoneOffset = tz->offset;
+    }
+    if (dst->validStarting <= utcTime.Value() && dst->offset != 0 && !dst->validUntil.IsNull())
+    {
+        dstOffset = dst->offset;
+    }
+    localTime.SetNonNull(utcTime.Value() + static_cast<uint64_t>(timeZoneOffset) + static_cast<uint64_t>(dstOffset));
+    LocalTime::Set(ep, localTime);
```

Comment:
Is this going to be running fairly often? If that's possible, it's probably worthwhile to calculate and set the absolute offset in another function with an expiry time. That way you can just check if the current offset is value, and you don't have to do any walking through the lists if recalculate. If it's expired, you can recalculate and re-balance the lists. (ditto for set commands)